File: src/kernel/Args.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/stream/IStringStream.h>
#include <base/stream/Serial.h>
#include <base/CmdArgs.h>
#include <base/Config.h>
#include <base/Machine.h>

#include "Args.h"

namespace kernel {

size_t Args::kmem          = 32 * 1024 * 1024;
cycles_t Args::timeslice   = 6000000;
const char *Args::fsimg    = nullptr;

void Args::usage(const char *name) {
    m3::Serial::get() << "Usage: " << name << " [-t=<timeslice>] [-f=<fsimg>] [-m=<kmem>] ...\n";
    m3::Serial::get() << "  -t: the timeslices for all VPEs\n";
    m3::Serial::get() << "  -f: the file system image (only used on host)\n";
    m3::Serial::get() << "  -m: the kernel memory size (> FIXED_KMEM)\n";
    m3::Machine::shutdown();
}

int Args::parse(int argc, char **argv) {
    int opt;
    while((opt = m3::CmdArgs::get(argc, argv, "f:t:m:")) != -1) {
        switch(opt) {
            case 'f': fsimg = m3::CmdArgs::arg; break;
            case 't': timeslice = m3::IStringStream::read_from<cycles_t>(m3::CmdArgs::arg); break;
            case 'm':
                kmem = m3::CmdArgs::to_size(m3::CmdArgs::arg);
                if(kmem <= FIXED_KMEM)
                    usage(argv[0]);
                break;
            default: usage(argv[0]);
        }
    }

    return m3::CmdArgs::ind;
}

}

File: src/kernel/src/args.rs
/*
 * Copyright (C) 2020-2021 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base::cell::{LazyStaticRefCell, Ref};
use base::cfg;
use base::col::Vec;
use base::env;

pub struct Args {
    pub kmem: usize,
    pub root_eps: usize,
}

impl Default for Args {
    fn default() -> Self {
        Self {
            kmem: 64 * 1024 * 1024,
            root_eps: cfg::DEF_EP_COUNT,
        }
    }
}

static ARGS: LazyStaticRefCell<Args> = LazyStaticRefCell::default();

pub fn get() -> Ref<'static, Args> {
    ARGS.borrow()
}

pub fn parse() {
    let mut args = Args::default();

    let get_size_arg = |argv: &Vec<&str>, i: &mut usize| -> usize {
        let size_str = argv.get(*i + 1).unwrap_or_else(|| usage());
        let size = parse_size(size_str).unwrap_or_else(|| usage());
        *i += 1;
        size
    };

    let mut i = 1;
    let argv: Vec<&str> = env::args().collect();
    while i < argv.len() {
        if argv[i] == "-m" {
            let kmem = get_size_arg(&argv, &mut i);
            if kmem <= cfg::FIXED_KMEM {
                usage();
            }
            args.kmem = kmem;
        }
        else if argv[i] == "-r:eps" {
            let ep_count = get_size_arg(&argv, &mut i);
            if ep_count == 0 {
                usage();
            }
            args.root_eps = ep_count;
        }
        i += 1;
    }

    ARGS.set(args);
}

fn usage() -> ! {
    panic!(
        "\nUsage: {} [-m <kmem>]
          -m: the kernel memory size (> FIXED_KMEM)",
        env::args().next().unwrap()
    );
}

fn parse_size(s: &str) -> Option<usize> {
    let mul = match s.chars().last() {
        Some(c) if c >= '0' && c <= '9' => 1,
        Some('k') | Some('K') => 1024,
        Some('m') | Some('M') => 1024 * 1024,
        Some('g') | Some('G') => 1024 * 1024 * 1024,
        _ => return None,
    };
    Some(match mul {
        1 => s.parse::<u64>().ok()? as usize,
        m => m * s[0..s.len() - 1].parse::<u64>().ok()? as usize,
    })
}
